### PR TITLE
Add demultiplexing subworkflow 10x

### PR DESCRIPTION
**Pull request by nschcolnicov**
This ticket addresses this issue:
https://github.com/nf-core/modules/pull/5984

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/demultiplex/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/demultiplex _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: conf/test_mkfastq.config [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    config_profile_name        = 'Test mkfastq profile'
+    config_profile_description = 'Minimal test dataset to check pipeline function'
+
+    // Limit resources so that this can run on GitHub Actions
+    max_cpus   = 2
+    max_memory = '8.GB'
+    max_time   = '1.h'
+
+    // Input data
+    input = 'https://raw.githubusercontent.com/nschcolnicov/test-datasets/demultiplex/samplesheet/1.3.0/mkfastq-samplesheet.csv' //TODO update to nf/core repo after pushing, also update samplesheet within samplesheet in test-datasets repo
```
Comment (nschcolnicov) [inline conf/test_mkfastq.config:23]: @fmalmeida @atrigila I will update this after everyone had a change to test it

File: README.md [hunk 2]
```diff
@@ -25,6 +25,7 @@
 2. Element Biosciences (via `bases2fastq`)
 3. Singular Genomics (via [`sgdemux`](https://github.com/Singular-Genomics/singular-demux))
 4. FASTQ files with user supplied read structures (via [`fqtk`](https://github.com/fulcrumgenomics/fqtk))
+5. 10x Genomics (via [`mkfastq`])
```
[the remark this answers is not in the document]
Comment (nschcolnicov) [reply README.md:28]: Done

File: conf/test_mkfastq.config [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
```
[the remark this answers is not in the document]
Comment (nschcolnicov) [reply conf/test_mkfastq.config]: Done thanks!

File: docs/output.md [hunk 4]
```diff
@@ -101,6 +102,27 @@ The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes d
...
+<summary>Output files</summary>
+
+| File                     | Directory                          | Description                                                  |
+| :----------------------- | :--------------------------------- | :----------------------------------------------------------- |
+| FASTQ                    | <OUTDIR>/<id>                      | Demultiplexed fastq.gz files                                 |
+| `*.fastp.html`           | <OUTDIR>/<id>                      | HTML report for fastp                                        |
+| `*.fastp.json`           | <OUTDIR>/<id>                      | JSON report for fastp                                        |
+| `*_summary.txt`          | <OUTDIR>/<id>                      | Summary statistics for the sequencing run                    |
+| Fastqc summary stats     | <OUTDIR>/<id>/\*fastqc_data.txt    | Per base quality summary, for each demultiplexed fastq file  |
+| Fastq summary html       | <OUTDIR>/<id>/\*fastqc_report.html | Interactive html link for fastqc summary stats               |
```
[the remark this answers is not in the document]
Comment (nschcolnicov) [reply docs/output.md]: Done thanks!

File: docs/output.md [hunk 5]
```diff
@@ -101,6 +102,27 @@ The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes d
 
 </details>
 
+### mkfastq
+
+[mkfastq](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/mkfastq) A tool for converting bcl files to fastq and demultiplexing for single-cell sequencing data.
```
[the remark this answers is not in the document]
Comment (nschcolnicov) [reply docs/output.md]: Done thanks!

File: modules/nf-core/cellranger/mkfastq/main.nf [hunk 6]
```diff
@@ -0,0 +1,74 @@
+process CELLRANGER_MKFASTQ {
+    tag {"$meta.lane" ? "$meta.id"+"."+"$meta.lane" : "$meta.id" }
+    label 'process_medium'
+
+    container "nf-core/cellrangermkfastq:8.0.0"
+
+    input:
+    tuple val(meta), path(bcl)
+    tuple val(meta), path(csv)
```
[the remark this answers is not in the document]
Comment (nschcolnicov) [reply modules/nf-core/cellranger/mkfastq/main.nf]: Done thanks!

File: subworkflows/local/mkfastq_demultiplex/main.nf [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+//
+// Demultiplex Singular Genomics bases data using mkfastq
+//
+
+include { CELLRANGER_MKFASTQ } from '../../../modules/nf-core/cellranger/mkfastq/main' 
+
+workflow MKFASTQ_DEMULTIPLEX {
+    take:
+        // Input fastq's must be bgzipped for compatibility with sgdemux
```
[the remark this answers is not in the document]
Comment (nschcolnicov) [reply subworkflows/local/mkfastq_demultiplex/main.nf]: Removed it, thanks!

File: subworkflows/local/mkfastq_demultiplex/main.nf [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        ch_mkfastq_input = ch_flowcell
+            .multiMap { meta, samplesheet, flowcell ->
+                    csv: [ meta, samplesheet ]
+                    bcl: [ meta, flowcell ]
+            }
+
+        // MODULE: mkfastq
+        CELLRANGER_MKFASTQ( ch_mkfastq_input.bcl, ch_mkfastq_input.csv )
+
+        // Generate meta for each fastq
```
[the remark this answers is not in the document]
Comment (nschcolnicov) [reply subworkflows/local/mkfastq_demultiplex/main.nf]: Done thanks!

File: docs/output.md [hunk 9]
```diff
@@ -101,6 +102,27 @@ The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes d
...
+<details markdown="1">
+<summary>Output files</summary>
+
+| File                     | Directory                          | Description                                                  |
+| :----------------------- | :--------------------------------- | :----------------------------------------------------------- |
+| FASTQ                    | <OUTDIR>/<id>                      | Demultiplexed fastq.gz files                                 |
+| `*.fastp.html`           | <OUTDIR>/<id>                      | HTML report for fastp                                        |
+| `*.fastp.json`           | <OUTDIR>/<id>                      | JSON report for fastp                                        |
+| `*_summary.txt`          | <OUTDIR>/<id>                      | Summary statistics for the sequencing run                    |
+| Fastqc summary stats     | <OUTDIR>/<id>/\*fastqc_data.txt    | Per base quality summary, for each demultiplexed fastq file  |
```
[the remark this answers is not in the document]
Comment (nschcolnicov) [reply docs/output.md]: Done thanks!

**Comment by nschcolnicov** [comment]
@nf-core-bot fix linting pretty please

File: tests/pipeline/mkfastq.nf.test.snap
Comment (fmalmeida) [inline tests/pipeline/mkfastq.nf.test.snap:1]: This snapshot seems to be missing the snaps of the filepaths.

File: workflows/demultiplex.nf [hunk 10]
```diff
@@ -230,7 +238,7 @@ workflow DEMULTIPLEX {
             )
         )
 
-        MULTIQC (
+        MULTIQC ( //TODO fix multiqc not resuming
```
Comment (fmalmeida) [inline workflows/demultiplex.nf:241]: Probably the summary maps used to collect run information might have some float variable like run name, star time, or something like this.

Would recommend first looking at the "ch_workflow_summary"

Comment (nschcolnicov) [reply workflows/demultiplex.nf:241]: Sounds good, I will address this on a separate ticket to keep things clean.

File: tests/pipeline/mkfastq.nf.test.snap
Comment (nschcolnicov) [reply tests/pipeline/mkfastq.nf.test.snap:1]: Nice catch, weird, Ill look into it.

**Comment by nschcolnicov** [comment]
The issue with the mkfastq CI test, seems to be unrelated to the scope of this ticket. I will remove it from the PR and address it in a separate ticket: https://github.com/nf-core/demultiplex/issues/211

**Pull request merged**